### PR TITLE
test: revisit tests for Accountants

### PR DESCRIPTION
**Pull request by odyslam**
- Adds a couple of fuzzing and unit tests
- Break large unit tests into more specific scenarios

### CODE REVIEW

**Comment by prestwich** [comment]
needs rebase to main and fixing

File: packages/contracts-bridge/contracts/test/NFTAccountant.t.sol [hunk 1]
```diff
@@ -85,41 +143,120 @@ contract NFTAccountantTest is Test {
         assertEq(accountant.totalMinted(asset), _prevTotal + amnt);
     }
 
-    function test_recordCheck() public {
+    function recordCheck(
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/test/NFTAccountant.t.sol:146]: this is all duplicated code with function above. de-duplicate by having the no-argument version invoke this version?

Comment (prestwich) [reply packages/contracts-bridge/contracts/test/NFTAccountant.t.sol:146]: both of these functions read local state in an unpleasant way that makes them dependent on the context in which they're invoked. Try to remove deps on test state by converting to argument?

E.g. these tests are very unclear, and brittle to changes in `recordCheck` and in local state

Screenshot 2022-11-21 at 9 36 14 AM

File: packages/contracts-bridge/contracts/test/NFTAccountant.t.sol [hunk 2]
```diff
@@ -85,41 +143,120 @@ contract NFTAccountantTest is Test {
...
+        address fuser,
+        uint256 famount
+    ) public {
+        vm.assume(!accountant.isAffectedAsset(fasset));
+        // Fuzz only for addresses that can in fact receive an ERC721
+        // Filters through the address of the test contract, VM, and others.
+        //
+        // Create2Deployer
+        vm.assume(fuser != 0x4e59b44847b379578588920cA78FbF26c0B4956C);
+        (bool success, bytes memory data) = fuser.call(
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/test/NFTAccountant.t.sol:239]: data is unused

File: packages/contracts-bridge/contracts/test/NFTAccountant.t.sol [hunk 3]
```diff
@@ -85,41 +143,120 @@ contract NFTAccountantTest is Test {
...
+    function testFuzz_recordSuccess(
+        uint8 assetIndex,
+        address payable user,
+        uint256 amount,
+        uint8 times
+    ) public {
+        // Fuzzing parameters
+        vm.assume(user != address(0));
+        assetIndex = uint8(bound(assetIndex, 0, 13));
+        address payable asset = accountant.affectedAssets()[assetIndex];
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/test/NFTAccountant.t.sol:187]: these lines are repeated 3 times. util method?

File: packages/contracts-bridge/contracts/test/NFTRecoveryAccountant.t.sol [hunk 4]
```diff
@@ -148,24 +163,86 @@ contract NFTRecoveryAccountantTest is Test {
...
+        // Total collected MUST be between 1 and totalAffected (inclusive)
+        uint256 collected = bound(collectedAssets, 1, totalAffected);
+        collectHelper(collected);
+        // Keep track of how many have been bridged back (instead of the circulation)
+        uint256 bridgedSum;
+        for (uint256 i; i < 8; i++) {
+            // Make sure that the user is not the address(0)
+            if (users[i] == address(0)) users[i] = address(0xBEEF);
+            // Make sure that the user can receive the NFT
+            (bool success, bytes memory data) = users[i].call(
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/test/NFTRecoveryAccountant.t.sol:203]: util method?

File: packages/contracts-bridge/contracts/test/NFTRecoveryAccountant.t.sol [hunk 5]
```diff
@@ -148,24 +163,86 @@ contract NFTRecoveryAccountantTest is Test {
...
+                vm.expectRevert("currently fully recovered");
+                vm.prank(fuser);
+                accountant.exposed_recover(id);
+                return;
+            }
+            recoverCheck(id);
             assertEq(
                 accountant.totalRecovered(asset),
-                _prevtotalRecovered + _recoverable
+                prevtotalRecovered + recoverable
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/test/NFTRecoveryAccountant.t.sol:245]: camel case

File: packages/contracts-bridge/contracts/test/NFTAccountant.t.sol [hunk 6]
```diff
@@ -10,18 +10,18 @@ import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
+    uint256 defaultAmount = 32000;
     ERC20Mock mockToken;
     NFTRecoveryAccountantHarness accountant;
 
     uint256 public constant AFFECTED_TOKEN_AMOUNT = 100_000_000;
 
     event ProcessFailure(
         uint256 indexed id,
-        address indexed asset,
+        address indexed defaultAsset,
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/test/NFTAccountant.t.sol:24]: artifact of find+replace ?

Comment (odyslam) [reply packages/contracts-bridge/contracts/test/NFTAccountant.t.sol:24]: ah

File: packages/contracts-bridge/contracts/test/NFTAccountant.t.sol [hunk 7]
```diff
@@ -30,18 +30,77 @@ contract NFTAccountantTest is Test {
         accountant = new NFTRecoveryAccountantHarness(address(this), recipient);
         accountant.initialize();
         // setup test vars
-        user = vm.addr(90210);
+        defaultUser = vm.addr(90210);
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/test/NFTAccountant.t.sol:33]: can we pre-emptively separate fixtures from mock state in prep for porting these to fork test?

File: packages/contracts-bridge/contracts/test/NFTAccountant.t.sol [hunk 8]
```diff
@@ -52,81 +111,171 @@ contract NFTAccountantTest is Test {
...
         accountant.ownerOf(_id);
         // calling record emits event
         vm.expectEmit(true, true, true, true, address(accountant));
-        emit ProcessFailure(_id, asset, user, amnt);
-        accountant.record(asset, user, amnt);
-        // next NFT is minted to user
-        assertEq(accountant.ownerOf(_id), user);
+        emit ProcessFailure(_id, _asset, _user, _amount);
+        accountant.record(_asset, _user, _amount);
+        // next NFT is minted to defaultUser
```
Comment (prestwich) [inline packages/contracts-bridge/contracts/test/NFTAccountant.t.sol:140]: find + replace artifact

File: packages/contracts-bridge/contracts/test/NFTRecoveryAccountant.t.sol (hunk 4 above)
Comment (prestwich) [reply packages/contracts-bridge/contracts/test/NFTRecoveryAccountant.t.sol:203]: util method was created ,but not used here

**Review by prestwich — APPROVED** [review]

**Pull request merged**